src/Storages/System/StorageSystemTables.cpp

```diff
@@ -74,20 +74,55 @@ namespace
 
 ColumnPtr getFilteredDatabases(const SelectQueryInfo & query_info, ContextPtr context)
 {
-    MutableColumnPtr column = ColumnString::create();
+    /** Engine in ststem.tables actually represents storage name. Some types of databases have their
+      * engine name same as their storage name, and it is time-consuming to connect to these kind of
+      * databases without a valid URL. So we use the engine name (storgae name) to avoid access to
+      * these kind of databases as much as possible. For example, when engine != 'PostgreSQL', we
+      * don't need to connect to the corresponding postgres database.
+      */
+    MutableColumnPtr database_column = ColumnString::create();
+    MutableColumnPtr engine_column = ColumnString::create();
 
     const auto databases = DatabaseCatalog::instance().getDatabases();
-    for (const auto & database_name : databases | boost::adaptors::map_keys)
+    for (const auto & [database_name, database] : databases)
     {
         if (database_name == DatabaseCatalog::TEMPORARY_DATABASE)
             continue; /// We don't want to show the internal database for temporary tables in system.tables
 
-        column->insert(database_name);
+        database_column->insert(database_name);
+        engine_column->insert(database->getEngineName());
     }
 
-    Block block { ColumnWithTypeAndName(std::move(column), std::make_shared<DataTypeString>(), "database") };
+    Block block
+    {
+        ColumnWithTypeAndName(std::move(database_column), std::make_shared<DataTypeString>(), "database"),
+        ColumnWithTypeAndName(std::move(engine_column), std::make_shared<DataTypeString>(), "engine")
+    };
     VirtualColumnUtils::filterBlockWithQuery(query_info.query, block, context);
-    return block.getByPosition(0).column;
+
+    /** All the hassle below is to make sure the special kind of database has been indeed filtered
+      * out when requested by user, but not incorrectly filter out other databases, because queries
+      * like engine = 'View', whose storage might be 'atomic' would be filtered out as well as its
+      * engine name != storage name. Filter for other databases happens at the execution stage.
+      */
```

Comment:
The fact that something could be filtered incorrectly means that the approach is not so good. I think we can do differently. We can add new methods to `IDatabase` interface that will help us to understand what databases can contain tables with specific engine. For example, 2 new methods:
```
virtual bool canContainOnlyOneTableEngine() { false; }
String getTablesEngineName() { throw Excpeption(ErrorCodes::LOGICAL_ERROR, "Not implemented") }
```

And implement these methods for databases PostgreSQL and MySQL (maybe some other, need to check). And when we will iterate through databases here we can add table engine to `engine_column` only when `canContainOnlyOneTableEngine->canContainOnlyOneTableEngine()` is true. In this case we will filter out only necessary databases.